packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/ContiguousScopeHandler.ts

```diff
@@ -0,0 +1,191 @@
+  }
+
+  generateScopeCandidates(
+    editor: TextEditor,
+    position: Position,
+    direction: Direction,
+  ): Iterable<TargetScope> {
+    return direction === "backward"
+      ? this.generateScopeCandidatesBackward(editor, position)
+      : this.generateScopeCandidatesForward(editor, position);
+  }
+
+  *generateScopeCandidatesBackward(
+    editor: TextEditor,
+    position: Position,
+  ): Iterable<TargetScope> {
+    let targetsForward = next(
+      getTargetsInDirection(this.scopeHandler, editor, position, "forward"),
+    );
+
+    const targetsBackward = getTargetsInDirection(
+      this.scopeHandler,
+      editor,
+      position,
+      "backward",
+    );
+
+    for (const targets of targetsBackward) {
+      if (targetsForward != null && isAdjacent(targets[1], targetsForward[0])) {
+        yield targetsToScope(targets[0], targetsForward[1]);
+        targetsForward = undefined;
+      } else {
+        yield targetsToScope(...targets);
+      }
+    }
+  }
+
+  *generateScopeCandidatesForward(
+    editor: TextEditor,
+    position: Position,
+  ): Iterable<TargetScope> {
+    let targetsBackward = next(
+      getTargetsInDirection(this.scopeHandler, editor, position, "backward"),
+    );
+
+    const targetsForward = getTargetsInDirection(
+      this.scopeHandler,
+      editor,
+      position,
+      "forward",
+    );
+
+    for (const targets of targetsForward) {
+      if (
+        targetsBackward != null &&
+        isAdjacent(targetsBackward[1], targets[0])
+      ) {
+        yield targetsToScope(targetsBackward[0], targets[1]);
+        targetsBackward = undefined;
+      } else {
+        yield targetsToScope(...targets);
+      }
+    }
+  }
+}
+
+function targetsToScope(
+  leadingTarget: Target,
+  trailingTarget: Target,
+): TargetScope {
+  if (leadingTarget.contentRange.isRangeEqual(trailingTarget.contentRange)) {
+    return {
+      editor: leadingTarget.editor,
+      domain: leadingTarget.contentRange,
+      getTargets: () => [leadingTarget],
+    };
+  }
+
+  const range = leadingTarget.contentRange.union(trailingTarget.contentRange);
+  return {
+    editor: leadingTarget.editor,
+    domain: range,
+    getTargets: () => [leadingTarget.withContentRange(range)],
+  };
+}
+
+function* getTargetsInDirection(
+  scopeHandler: ScopeHandler,
+  editor: TextEditor,
+  position: Position,
+  direction: Direction,
+): Iterable<[Target, Target]> {
+  const isForward = direction === "forward";
+  let first, last: Target | undefined;
+
+  const generator = scopeHandler.generateScopes(editor, position, direction, {
+    allowAdjacentScopes: true,
+    skipAncestorScopes: true,
+  });
+
+  for (const scope of generator) {
+    for (const target of scope.getTargets(false)) {
+      if (first == null) {
+        first = target;
+      }
+
+      if (last != null) {
+        const [leadingTarget, trailingTarget] = isForward
+          ? [last, target]
+          : [target, last];
```

Comment:
Tempted to switch to a `{proximal, distal}` obj to remove all these ternaries and confusing anonymous lists

Reply:
Switched to proximal and distal